kore/src/Kore/Logger.hs

```diff
@@ -240,18 +295,82 @@ withLogScope newScope = localLogAction (contramap appendScope)
+    toEntry :: entry -> SomeEntry
+    toEntry = SomeEntry
+
+    fromEntry :: SomeEntry -> Maybe entry
+    fromEntry (SomeEntry entry) = Data.Typeable.cast entry
+
+    shouldLog :: Severity -> Set Scope -> entry -> Bool
+
+data SomeEntry where
+    SomeEntry :: Entry entry => entry -> SomeEntry
+
+instance Pretty.Pretty SomeEntry where
+    pretty (SomeEntry entry) = Pretty.pretty entry
+
+withSomeEntry
+    :: (forall entry. Entry entry => entry -> a)
+    -> SomeEntry
+    -> a
+withSomeEntry f (SomeEntry entry) = f entry
+
+instance Entry LogMessage where
+    shouldLog :: Severity -> Set Scope -> LogMessage -> Bool
+    shouldLog minSeverity currentScope LogMessage { severity, scope } =
+        severity >= minSeverity && scope `member` currentScope
+     where
+       member :: [Scope] -> Set Scope -> Bool
+       member s set
+         | Set.null set = True
+         | otherwise    = Fold.any (`Set.member` set) s
+
+class Monad m => MonadLog m where
+    logM :: Entry entry => entry -> m ()
+    logScope :: Entry e1 => Entry e2 => (e1 -> e2) -> m a -> m a
+
 newtype LoggerT m a =
-    LoggerT { getLoggerT :: ReaderT (LogAction m LogMessage) m a }
+    LoggerT { getLoggerT :: ReaderT (LogAction m SomeEntry) m a }
     deriving (Functor, Applicative, Monad)
     deriving (MonadIO)
 
-instance MonadTrans LoggerT where
-    lift = LoggerT . Monad.Trans.lift
-    {-# INLINE lift #-}
+instance Monad m => MonadLog (LoggerT m) where
+    logM entry =
+        LoggerT $ ask >>= Monad.Trans.lift . (<& toEntry entry)
+    logScope f (LoggerT (ReaderT logActionReader)) =
+        LoggerT . ReaderT
+            $ \(LogAction logAction) ->
+                logActionReader . LogAction
+                    $ \entry ->
+                        case fromEntry entry of
+                            Nothing -> logAction entry
+                            Just entry' -> logAction $ toEntry $ f entry'
 
 instance Monad m => WithLog LogMessage (LoggerT m) where
-    askLogAction = LoggerT . fmap liftLogAction $ ask
+    askLogAction :: LoggerT m (LogAction (LoggerT m) LogMessage)
+    askLogAction = LoggerT . ReaderT $ pure . go
+      where
+        go :: LogAction m SomeEntry -> LogAction (LoggerT m) LogMessage
+        go (LogAction fn) = LogAction $ Monad.Trans.lift . fn . toEntry
     {-# INLINE askLogAction #-}
 
-    localLogAction locally = LoggerT . local locally . getLoggerT
+    localLogAction
+        :: (forall n. LogAction n LogMessage -> LogAction n LogMessage)
+        -> LoggerT m a
+        -> LoggerT m a
+    localLogAction locally (LoggerT (ReaderT la)) =
+        LoggerT . ReaderT $ la . go locally
+      where
+        go
+            :: (LogAction n LogMessage -> LogAction n LogMessage)
+            -> (LogAction n SomeEntry -> LogAction n SomeEntry)
+        go =
+            Profunctor.dimap
+                (contramap toEntry)
+                (contramap (fromJust . fromEntry))
```

Comment:
Can this `fromJust` ever fail?

Reply:
I don't think so, we're going from `LogMessage` to `LogMessage`, so it should always work.

Reply:
No, that can definitely fail. I don't think you can write this using `Profunctor`. If it helps, you can think of the helper as having the signature
```.hs
go
    :: (LogAction n LogMessage -> LogAction n LogMessage)
    -> LogAction n SomeEntry
    -> LogAction n SomeEntry
```
In other words, you can match on the `LogAction n SomeEntry` and choose to apply the first argument, or not. I think you need a `Choice` to do it the profunctor way, but I'd just use ordinary functions here.